src/kernels/debugger/debuggingManagerBase.ts

```diff
@@ -0,0 +1,206 @@
+import { DebuggingTelemetry } from './constants';
+import { sendTelemetryEvent } from '../../telemetry';
+import { traceError, traceInfoIfCI } from '../../platform/logging';
+import { DataScience } from '../../platform/common/utils/localize';
+import { IKernelDebugAdapterConfig } from './types';
+import { Debugger } from '../../notebooks/debugger/debugger';
+import { KernelDebugAdapterBase } from './kernelDebugAdapterBase';
+import { INotebookControllerManager } from '../../notebooks/types';
+import { IpykernelCheckResult, isUsingIpykernel6OrLater } from '../../notebooks/debugger/helper';
+
+/**
+ * The DebuggingManager maintains the mapping between notebook documents and debug sessions.
+ */
+export abstract class DebuggingManagerBase implements IDisposable {
+    private notebookToDebugger = new Map<NotebookDocument, Debugger>();
+    protected notebookToDebugAdapter = new Map<NotebookDocument, KernelDebugAdapterBase>();
+    protected notebookInProgress = new Set<NotebookDocument>();
+    protected readonly disposables: IDisposable[] = [];
+    private _doneDebugging = new EventEmitter<void>();
+
+    public constructor(
+        private kernelProvider: IKernelProvider,
+        private readonly notebookControllerManager: INotebookControllerManager,
+        protected readonly commandManager: ICommandManager,
+        protected readonly appShell: IApplicationShell,
+        protected readonly vscNotebook: IVSCodeNotebook
+    ) {}
+
+    public async activate() {
+        this.disposables.push(
+            // track termination of debug sessions
+            debug.onDidTerminateDebugSession(this.endSession.bind(this)),
+
+            // track closing of notebooks documents
+            workspace.onDidCloseNotebookDocument(async (document) => {
+                const dbg = this.notebookToDebugger.get(document);
+                if (dbg) {
+                    await dbg.stop();
+                    this.onDidStopDebugging(document);
+                }
+            })
+        );
+    }
+    public getDebugCell(notebook: NotebookDocument): NotebookCell | undefined {
+        return this.notebookToDebugAdapter.get(notebook)?.debugCell;
+    }
+
+    public get onDoneDebugging(): Event<void> {
+        return this._doneDebugging.event;
+    }
+
+    public dispose() {
+        this.disposables.forEach((d) => d.dispose());
+    }
+
+    public isDebugging(notebook: NotebookDocument): boolean {
+        return this.notebookToDebugger.has(notebook);
+    }
+
+    public getDebugSession(notebook: NotebookDocument): Promise<DebugSession> | undefined {
+        const dbg = this.notebookToDebugger.get(notebook);
+        if (dbg) {
+            return dbg.session;
+        }
+    }
+    public getDebugAdapter(notebook: NotebookDocument): KernelDebugAdapterBase | undefined {
+        return this.notebookToDebugAdapter.get(notebook);
+    }
+
+    protected onDidStopDebugging(_notebook: NotebookDocument): void {
```

Comment:
abstract?

Reply:
No, it should not have to be overridden, hence didn't go with abstract (ie. overriding is optional)